File: backend/src/main/java/com/eitek/incidents/domain/models/Assignee.java
package com.eitek.incidents.domain.models;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

import org.checkerframework.checker.units.qual.A;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Assignee {
    private Long id;
    private String name;
    private String role;
    private boolean isActive;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    
    public void deactivate() {
        this.isActive = false;
        this.updatedAt = LocalDateTime.now();
    }

    public void activate() {
        this.isActive = true;
        this.updatedAt = LocalDateTime.now();
    }
    
    /**
     * Verifica si este assignee puede asignar incidentes a otros
     * Solo admin y support tienen permisos de asignación
     */
    public boolean canAssignIncidents() {
        return AssigneeRoles.ADMIN.name().equalsIgnoreCase(this.role) ||
               AssigneeRoles.SUPPORT.name().equalsIgnoreCase(this.role);
    }
    
    /**
     * Verifica si este assignee puede asignar a un rol específico
     */
    public boolean canAssignToRole(String targetRole) {
        if (!canAssignIncidents()) {
            return false; // Sin permisos de asignación
        }
        return true; // admin y support pueden asignar a cualquier rol
    }
    
    /**
     * Verifica si este assignee puede eliminar incidentes
     * Solo admin tiene permisos de eliminación
     */
    public boolean canDeleteIncidents() {
        return AssigneeRoles.ADMIN.name().equalsIgnoreCase(this.role) ||
               AssigneeRoles.SUPPORT.name().equalsIgnoreCase(this.role);
    }

    public boolean isAdmin() {
        return AssigneeRoles.ADMIN.name().equalsIgnoreCase(this.role);
    }
}

File: frontend/src/core/domain/Assignee.ts
import { AssigneeRoles } from "./AssigneeRoles"

export class Assignee {
    id: number
    name: string
    role: AssigneeRoles
    isActive: boolean
    initials: string
    createdAt: Date
    updatedAt: Date

    constructor({
        id,
        name,
        role,
        isActive,
        createdAt,
        updatedAt
    }: {
        id: number
        name: string
        role: AssigneeRoles
        isActive: boolean
        createdAt: Date
        updatedAt: Date
    }) {
        this.id = id
        this.name = name || ""
        this.role = role || AssigneeRoles.USER
        this.isActive = isActive || true
        this.createdAt = createdAt || new Date()
        this.updatedAt = updatedAt || new Date()
        this.initials = this.computeInitials()
    }

    canAssignIncidents(): boolean {
        return this.role === AssigneeRoles.ADMIN || this.role === AssigneeRoles.SUPPORT;
    }

    canAssignToRole(_targetRole: AssigneeRoles): boolean {
        if (!this.canAssignIncidents()) {
            return false;
        }
        return true;
    }

    canDeleteIncidents(): boolean {
        return this.role === AssigneeRoles.ADMIN;
    }

    computeInitials(): string {
        const names = this.name.trim().split(" ");
        if (names.length === 0) return "";
        if (names.length === 1) return names[0].charAt(0).toUpperCase();
        return (names[0].charAt(0) + names[names.length - 1].charAt(0)).toUpperCase();
    }

    activate(): void {
        this.isActive = true;
        this.updatedAt = new Date();
    }

    deactivate(): void {
        this.isActive = false;
        this.updatedAt = new Date();
    }
}